GPU stats

Raalsky opened the pull request:
@grzegorz-gutowski

Raalsky committed:
Initial

Raalsky committed:
Merge branch 'memory-limiter' into gpu-stats

Raalsky committed:
Added gpu summary memory utilization

grzegorz-gutowski commented:
Few things:
1. We use gpu_memory limit, as limit per card. I would like to keep that.
2. To keep it compatible, we should use gpu_memory, as maximum usage per card.
3. If we impose gpu_memory limit, we should decrease the gpu_memory stat by the ammount of memory reserved by us.
4.Do we need some safe checks for systems without nvidia driver?

Raalsky committed:
Max insted of sum

Raalsky committed:
Explicitly called memory reservation instead of subprocess

Raalsky commented:
> Few things:
> 
>     1. We use gpu_memory limit, as limit per card. I would like to keep that.
> 
>     2. To keep it compatible, we should use gpu_memory, as maximum usage per card.
> 
>     3. If we impose gpu_memory limit, we should decrease the gpu_memory stat by the ammount of memory reserved by us.
>        4.Do we need some safe checks for systems without nvidia driver?

Ad 1 & 2. Sure, fixed
Ad 3. Check [my latest commit](https://github.com/Raalsky/kolejka-judge/pull/2/commits/96776b281f1bcb8cc274617a43ef20bdb9537e4a) . It was totally buggy due to the fact that memory reservation process needs some time to allocate memory. So our "main" command was executed with some memory allocation and stats were correctly computed before the preserving subprocess allocates data. Right now I've placed memory reservation as a part of LocalSystem so main command will not be executed before the memory limitation. The logic of GPU memory usage is computed as `APPLIED LIMIT - (TOTAL MEMORY - CURRENT GPU MEMORY USAGE)`. As `TOTAL MEMORY - CURRENT MEMORY USAGE = FREE` and `APPLIED LIMIT - FREE = CURRENT PROCESS USAGE`.
Ad 4. I've added checking for limit existence in the monitor process so the logic is placed in memory preservation methods.

Sorry for being a little bit distracted. I'm totally sick for a few days 🤒
